database.go

```diff
@@ -0,0 +1,272 @@
+				if index >= len(output) {
+					return offset, metrics, output
+				}
+
+				current := output[index]
+				// if the current metric exists already, simply increment it and continue
+				if current.Value() == metric.Value() {
+					totalLength = totalLength + metric.Count()
+					current.IncrBy(metric.Count())
+					offset = offset + metric.Count()
+					index = index + 1
+					break
+				} else if current.Value() > metric.Value() {
+					// our metric should go in front of the existing metric
+					// both lists are sorted, so this means that our metrics is "behind" the output
+					totalLength = totalLength + metric.Count()
+					offset = offset + metric.Count()
+
+					// there's an edge case where we need to create a new array if this is the first item
+					if index == 0 {
+						// prepend the item
+						output = append([]*BulkMetric{metric}, output...)
+					} else {
+						// inject the item at this place in the array
+						head := append(output[:index], metric)
+						output = append(head, output[index:]...)
+					}
+					break
+				} else {
+					// keep looking, the current metric is greater than the value we're at in the existing array
+					index = index + 1
+				}
+			}
+		}
+
+		// if we get to this level, it means that we've inserted all the metrics into the database
+		return offset, []*BulkMetric{}, output
+	}
+
+	recalculate := func() {
+		if len(left) == 0 {
+			return
+		}
+
+		leftTail := left[len(left)-1].Value()
+
+		// if total is odd then we grab the last item from the left array
+		if totalLength%2 == 1 {
+			m.median = int32(leftTail)
+			return
+		}
+
+		rightTail := right[0].Value()
+		median := (rightTail + leftTail) / 2
+		m.median = int32(median)
+	}
+
+	// take items from left and move them right until the two arrays are balanced!
+	rebalanceRight := func(l, r []*BulkMetric, offset int) ([]*BulkMetric, []*BulkMetric) {
+		for {
+			if offset < 1 || len(l) < 1 {
+				break
+			}
+
+			// if the head of right is greater than offset then we
+			// can't pop it completely so we split it between the two lists until they are equal
+			lTail := l[len(l)-1]
+			if lTail.Count() > offset {
+				// remove the offset from the rHead
+				lTail.DecrBy(offset)
+
+				// create a new tail for l
+				rHead := NewBulkMetric(lTail.Value())
+				// bulkMetric defaults to 1 so we need to account for that when setting its value
+				rHead.IncrBy(offset - 1)
+
+				// finally we add this as the new head of the r array
+				r = append([]*BulkMetric{rHead}, r...)
+				break
+			}
+
+			// the head of r is less than the offset, so we can pop the whole node off of r and put it on l
+			offset -= lTail.Count()
+			r = append([]*BulkMetric{lTail}, r...)
+			l = l[:len(l)-1]
+		}
+
+		return l, r
+	}
+
+	// take items from right and move them left until the two arrays are balanced!
+	rebalanceLeft := func(l, r []*BulkMetric, offset int) ([]*BulkMetric, []*BulkMetric) {
```

Comment:
Purposefully wrote out `rebalanceLeft` and `rebalanceRight` instead of trying to make it a single utility function; the index and indirection is out of hand.

**todo** come back to this, though.